nixos/syncthing: add guiPasswordFile option, add and move Syncthing tests

nullcubee opened the pull request:
Supercedes: #290485, #393239
Fixes: #85336

PR originally by @universenox, with requested changes from myself and @Enzime, plus the tests I made in #393239.

Added a guiPasswordFile option to the Syncthing module so that you can set the Syncthing GUI password from a file, i.e. using sops-nix.

Moved the Syncthing tests to their own folder.

Added two tests for setting the GUI password, one for setting a plaintext password, and one using the new guiPasswordFile option. These could in theory be combined into a single test with multiple nodes, but one test per method was easier, even though they do share most of their code. This could be changed if desired.

There is also a check to make sure the plain text version of the password doesn't appear in the final syncthing config but I'm not sure if that is strictly necessary, or if the check works properly. Not sure how to best test that. 

To check if the password succeeded I check to see if `This Device` appears on screen, which is a part of the normal Syncthing web interface when you're authenticated. I did notice as well the window title also changes from `(unknown device) | Syncthing` to the hostname of the machine, which is a faster check, but I'm assuming that checking for "This Device" would be more robust? You might also be able to curl or query the config of syncthing via commands instead, which might be a better approach, but I'm not sure how to set that up, and I wanted to check if the web interface was working properly with the password as well, so I'm just using the full xorg session for the test.

CC Syncthing maintainers:
@jokogr @peterhoeg 

CC Original PR Reviewers:
@Enzime @bjornfor @leona-ya 

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [x] [NixOS tests] in [nixos/tests].
  - [x] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nullcubee committed:
nixos/tests/syncthing: add guiPassword and guiPasswordFile tests

universenox committed:
nixos/syncthing: add guiPasswordFile option

nullcubee committed:
nixos/tests/syncthing: move tests to folder

bjornfor commented:
I checked out the PR and got:

```
$ nom-build -A syncthing.tests
[...]
       error: attribute 'gui' missing
       at /home/bf/nixpkgs/nixos/modules/services/networking/syncthing.nix:856:22:
          855|       {
          856|         assertion = (cfg.settings.gui ? password) -> cfg.guiPasswordFile == null;
             |                      ^
          857|         message = ''
```

nullcubee committed:
syncthing: add missing syncthing-folders test

bjornfor commented:
Eval issue is fixed, but `nix-build -A syncthing.tests` still fails for me.

nullcubee committed:
nixos/tests/syncthing: disable syncthing-many-devices test

The `syncthing-many-devices` test has been failing since at least
2025-07-06, and there doesn't appear to be any issue reports relating to
it. Disabling is done simply out of convenience, and an issue has been
reported about it.

Tracking issue: https://github.com/NixOS/nixpkgs/issues/447674
First failing hydra build: https://hydra.nixos.org/build/302090970

nullcubee commented:
It seems to get stuck on the `syncthing-many-devices` test for me, checking Hydra reveals that this test has been failing for a good couple months now: <https://hydra.nixos.org/build/308267277>, and the test failing has been mentioned before: <https://github.com/NixOS/nixpkgs/pull/436432#pullrequestreview-3157344506>, so that shouldn't affect this PR but it's worth mentioning. CC @doronbehar, you're listed as the maintainer of that test?

Anyways, commenting out `syncthing-many-devices` in the syncthing package's tests allows all the tests to pass successfully, so everything else seems to works fine. Maybe I should comment it out in this PR as well? Not sure.

bjornfor commented:
> Maybe I should comment it out in this PR as well? Not sure.

I have no strong opinion on which way we move forward, but I don't think we should merge another syncthing PR before making `nix-build -A syncthing.tests` pass. That might be solved by disabling the test or fixing it.

nullcubee commented:
I created #447674 to keep track of the failure with the test, and for now I'm going to just disable the test so we can move forward with this PR.

doronbehar commented on nixos/tests/all-tests.nix:
```diff
@@ -1404,12 +1404,12 @@ in
+  syncthing-init = runTest ./syncthing/syncthing-init.nix;
+  syncthing-many-devices = runTest ./syncthing/syncthing-many-devices.nix;
+  syncthing-no-settings = runTest ./syncthing/syncthing-no-settings.nix;
+  syncthing-relay = runTest ./syncthing/syncthing-relay.nix;
```
```suggestion
  syncthing = runTest ./syncthing/main.nix;
  syncthing-folders = runTest ./syncthing/folders.nix;
  syncthing-init = runTest ./syncthing/init.nix;
  syncthing-many-devices = runTest ./syncthing/many-devices.nix;
  syncthing-no-settings = runTest ./syncthing/no-settings.nix;
  syncthing-relay = runTest ./syncthing/relay.nix
```

doronbehar commented on nixos/modules/services/networking/syncthing.nix:
```diff
@@ -251,6 +252,12 @@ let
+    +
+      # Now we hash the contents of guiPasswordFile and use the result to update the gui password
+      (lib.optionalString (cfg.guiPasswordFile != null) ''
+        ${pkgs.mkpasswd}/bin/mkpasswd -m bcrypt --stdin <"${cfg.guiPasswordFile}" | tr -d "\n" > "$RUNTIME_DIRECTORY/password_bcrypt"
```
I am curious as to why is `tr` needed?

doronbehar commented on nixos/tests/all-tests.nix:
```diff
@@ -1406,6 +1406,8 @@ in
   sympa = runTest ./sympa.nix;
   syncthing = runTest ./syncthing/syncthing.nix;
   syncthing-folders = runTest ./syncthing/syncthing-folders.nix;
+  syncthing-guiPassword = runTest ./syncthing/syncthing-guiPassword.nix;
+  syncthing-guiPasswordFile = runTest ./syncthing/syncthing-guiPasswordFile.nix;
```
```suggestion
  syncthing-guiPassword = runTest ./syncthing/guiPassword.nix;
  syncthing-guiPasswordFile = runTest ./syncthing/guiPasswordFile.nix;
```

doronbehar commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -67,6 +67,9 @@ let
           inherit (nixosTests)
             syncthing
+            syncthing-folders
+            syncthing-guiPassword
+            syncthing-guiPasswordFile
```
Specifically these 2 can be squashed into the previous commit.

doronbehar commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -67,6 +67,9 @@ let
         tests = {
           inherit (nixosTests)
             syncthing
+            syncthing-folders
```
This one was missing from before and I'd put it in a separate commit.

doronbehar commented on pkgs/applications/networking/syncthing/default.nix:
```diff
@@ -67,6 +67,9 @@ let
+            syncthing-guiPassword
+            syncthing-guiPasswordFile
             syncthing-init
             syncthing-many-devices
```
This line should be commented out in the next commit.

doronbehar replied:
Oh I'm sorry I got confused a bit. I think it should be commented out in the `all-tests.nix` file, with a comment linking to the nixpkgs issue etc. The line here should be removed completely IMO, to not write the comment twice.

nullcubee replied:
ah okay, that sounds good to me!

doronbehar reviewed (commented):
Broadly speaking the changes look good. Should be good to go after these small comments are addressed.

doronbehar commented:
OK this should be good to go, but I'm putting a bit of effort now to maybe fix the disabled test...

doronbehar commented:
I'll take care of that later. That test _is_ important for me.
